### PR TITLE
create index, read index stream and create table with timeout index

### PR DESCRIPTION
**Pull request by badhrinathpa**

### CODE REVIEW

File: dbtestapp/dbtestapp.go [hunk 1]
```diff
@@ -61,7 +74,21 @@ func main() {
 		log.Println("Error getting student: " + err.Error())
 	}
 
-	createDocumentWithTimeout()
+	log.Println("starting timeout document")
+	database := MongoDBLibrary.Client.Database("sdcore")
+	timeoutColl := database.Collection("timeout")
+	//create stream to monitor actions on the collection
+	timeoutStream, err := timeoutColl.Watch(context.TODO(), mongo.Pipeline{})
```
Comment (thakurajayL) [inline dbtestapp/dbtestapp.go:81]: Can multiple PODs watch for same collection?

File: api_mongoDB.go [hunk 2]
```diff
@@ -379,12 +381,83 @@ func GetOneCustomDataStructure(collName string, filter bson.M) (bson.M, error) {
...
+		Options: options.Index().SetUnique(true),
+	}
+
+	idx, err := collection.Indexes().CreateOne(context.Background(), index)
+	if err != nil {
+		logger.MongoDBLog.Error("Create Index failed : ", keyField, err)
+		return false, err
+	}
+
+	logger.MongoDBLog.Println("index : ", idx)
```
Comment (thakurajayL) [inline api_mongoDB.go:417]: add key field in the logs as well.

File: api_mongoDB.go [hunk 3]
```diff
@@ -379,12 +381,83 @@ func GetOneCustomDataStructure(collName string, filter bson.M) (bson.M, error) {
...
+
+	var result []bson.M
+	if err = cursor.All(context.TODO(), &result); err != nil {
+		logger.MongoDBLog.Fatal(err)
+	}
+
+	for _, v := range result {
+		for k1, v1 := range v {
+			valStr := fmt.Sprint(v1)
+			if (k1 == "name") && strings.Contains(valStr, timeField) {
```
Comment (thakurajayL) [inline api_mongoDB.go:440]: can we have k1 field as parameter in the function ?

File: api_mongoDB.go [hunk 4]
```diff
@@ -379,12 +381,83 @@ func GetOneCustomDataStructure(collName string, filter bson.M) (bson.M, error) {
...
+		logger.MongoDBLog.Error("Create Index failed : ", keyField, err)
+		return false, err
+	}
+
+	logger.MongoDBLog.Println("index : ", idx)
+
+	return true, nil
+}
+
+func RestfulAPIPatchOneTimeout(collName string, filter bson.M, putData map[string]interface{}, timeout int32, timeField string) bool {
```
Comment (thakurajayL) [inline api_mongoDB.go:422]: Can you please help content on the API.
e.g. What happens if timeout updates.

Comment (thakurajayL) [reply api_mongoDB.go:422]: Current restriction is that all NFs to have same timeout ?

File: dbtestapp/dbtestapp.go [hunk 5]
```diff
@@ -6,35 +6,48 @@
...
+		log.Println("iterate stream %v\n", data)
+	}
 }
 
 func main() {
 	log.Println("dbtestapp started")
 
 	// connect to mongoDB
-	MongoDBLibrary.SetMongoDB("sdcore", "mongodb://mongodb:27017")
+	MongoDBLibrary.SetMongoDB("sdcore", "mongodb://mongodb-headless:27017")
```
Comment (thakurajayL) [inline dbtestapp/dbtestapp.go:43]: Can we take this in DBTestapp config ?

File: dbtestapp/dbtestapp.go [hunk 6]
```diff
@@ -6,35 +6,48 @@
...
+
+	_, errVal := MongoDBLibrary.CreateIndex("student", "Name")
+	if errVal != nil {
+		log.Println("Create index failed on Name field : ", errVal)
+	}
 
+	insertStudentInDB("Osman Amjad", 21)
```
Comment (thakurajayL) [inline dbtestapp/dbtestapp.go:50]: Add comment about findings

File: api_mongoDB.go [hunk 7]
```diff
@@ -10,8 +10,10 @@ import (
 	"context"
 	"encoding/json"
 	"errors"
+	"fmt"
```
Comment (thakurajayL) [inline api_mongoDB.go:13]: Can you add Intel Copyright in appropriate files ?

File: api_mongoDB.go (hunk 3 above)
Comment (badhrinathpa) [reply api_mongoDB.go:440]: The K1 field will be "name" as will be explained in the comments

File: dbtestapp/dbtestapp.go (hunk 1 above)
Comment (badhrinathpa) [reply dbtestapp/dbtestapp.go:85]: Yes they can. tested with multiple dbtestapp pods. But index create might fail for one of the pods cause both will try to create timeout index on same collection with same timeout value. But even if it fails, the one that succeeded would have created the index. so it should be ok.

File: api_mongoDB.go (hunk 2 above)
Comment (badhrinathpa) [reply api_mongoDB.go]: done

File: api_mongoDB.go (hunk 4 above)
Comment (badhrinathpa) [reply api_mongoDB.go:428]: done

File: dbtestapp/dbtestapp.go (hunk 5 above)
Comment (badhrinathpa) [reply dbtestapp/dbtestapp.go]: will do in next iteration

File: dbtestapp/dbtestapp.go (hunk 6 above)
Comment (badhrinathpa) [reply dbtestapp/dbtestapp.go]: done

File: api_mongoDB.go (hunk 7 above)
Comment (badhrinathpa) [reply api_mongoDB.go:14]: done

File: api_mongoDB.go [hunk 8]
```diff
@@ -379,12 +382,96 @@ func GetOneCustomDataStructure(collName string, filter bson.M) (bson.M, error) {
...
+	// loop through the map and check for entry with key as name
+	// for every entry with key as name,check if the value string contains the timeField string.
+	// the Indexes are generally named such as follows :
+	// field name : createdAt, index name : createdAt_1
+	// drop the index if found.
+	for _, v := range result {
+		for k1, v1 := range v {
+			valStr := fmt.Sprint(v1)
+			if (k1 == "name") && strings.Contains(valStr, timeField) {
+				_, err = collection.Indexes().DropOne(context.Background(), valStr)
```
Comment (badhrinathpa) [inline api_mongoDB.go:454]: break after value is found

File: api_mongoDB.go [hunk 9]
```diff
@@ -379,12 +382,96 @@ func GetOneCustomDataStructure(collName string, filter bson.M) (bson.M, error) {
...
+
+	//create new index with new timeout
+	index := mongo.IndexModel{
+		Keys:    bsonx.Doc{{Key: timeField, Value: bsonx.Int32(1)}},
+		Options: options.Index().SetExpireAfterSeconds(timeout),
+	}
+
+	_, err = collection.Indexes().CreateOne(context.Background(), index)
+	if err != nil {
+		logger.MongoDBLog.Panic(err)
```
Comment (badhrinathpa) [inline api_mongoDB.go:470]: return error instead of panic

File: dbtestapp/dbtestapp.go [hunk 10]
```diff
@@ -35,8 +43,17 @@ func main() {
 	// connect to mongoDB
 	MongoDBLibrary.SetMongoDB("sdcore", "mongodb://mongodb:27017")
 
-	insertStudentInDB("Osman Amjad", 21)
-	student, err := getStudentFromDB("Osman Amjad")
+	_, errVal := MongoDBLibrary.CreateIndex("student", "Name")
```
Comment (badhrinathpa) [inline dbtestapp/dbtestapp.go:46]: Set mongodb name through configuration.

File: api_mongoDB.go (hunk 8 above)
Comment (badhrinathpa) [reply api_mongoDB.go:457]: done

**Review by thakurajayL — APPROVED** [review]

**Pull request merged**